packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -1065,11 +1126,31 @@ function useMutableSource<Source, Snapshot>(
   //
   // In both cases, we need to throw away pending udpates (since they are no longer relevant)
   // and treat reading from the source as we do in the mount case.
+  const didGetSnapshotChange = !is(prevGetSnapshot, getSnapshot);
   if (
     !is(prevSource, source) ||
     !is(prevSubscribe, subscribe) ||
-    !is(prevGetSnapshot, getSnapshot)
+    didGetSnapshotChange
   ) {
+    if (didGetSnapshotChange) {
```

Comment:
Can remove this nested check. Need to drop updates from new sources, too.

Reply:
Yeah, I wasn't sure about that... I'll remove it :smile:

Thanks for the review!

Reply:
Hm, I think removing this check actually breaks some subtle assumptions about the effects and when we need to update the `getSnapshot`/`setSnapshot` refs. I think it would cause us to drop updates for mutations after a source changed... need to think about it a bit.

Reply:
I think it works because when `source` or `subscribe` changes, you're going to unsubscribe/resubscribe.

When `getSnapshot` changes, you don't resubscribe but you do track the latest value with a ref. Which I think of as an optimization for resubscribing; you're still "changing" the subscription in both cases. So they are the same thing except for what we do in the commit phase. (For example, I think all your tests would pass if you removed the effect and ref that tracks `getSnapshot` and instead added `getSnapshot` as a dependency to the  resubscribe hook.)

Reply:
Ah, I see how to fix it. Will let me simplify things a bit too...

Reply:
> For example, I think all your tests would pass if you removed the effect and ref that tracks `getSnapshot` and instead added `getSnapshot` as a dependency to the  resubscribe hook.

Just checked this to confirm and they do all pass

Reply:
Okay, here's what 86da5ce did:

Recreating the queue and dispatch if `source` or `subscribe` changed too (not just `getSnapshot`) meant that the we needed to update `refs.setSnapshot` in that case too. (Before that effect was only run when `getSnapshot` changed.)

So I added `source` or `subscribe` as dependencies to the first effect so it would update the `dispatch` ref. Since we're now *also* running this effect whenever we resubscribe, I was also able to consolidate the check for a changed snapshot between render and passive effects into just the first effect (rather than both).

Reply:
Weird. I didn't see any of your interleaved commits until just now, after I left mine. Heh. Didn't mean to seem like I was ignoring you.

> For example, I think all your tests would pass if you removed the effect and ref that tracks getSnapshot and instead added getSnapshot as a dependency to the resubscribe hook.

This requires us to over-subscribe though, (since we can't share that effect without running the previous cleanup function and *unsubscribing*). Once we replace `useEffect` with our own custom pushed effect, we can optimize this though.

Reply:
To be clear, I'm trying to avoid re-subscribing unless *actually necessary* (not just convenient) because subscriptions can be expensive for some of `useMutableSource`'s users.

Reply:
Right, I can see why you did it this way. I think we might want to revisit the trade off in the future, though, since changing `getSnapshot` is expensive regardless because it leads to deopts. So it needs to be relatively stable. (I am kinda worried about this, since it's easy for users to mess up.) And if we assume it's relatively stable, then resubscribing might not be so bad.

Reply:
> since changing getSnapshot is expensive regardless because it leads to deopts. So it needs to be relatively stable.

Yeah, absolutely. Maybe there's something we could do to detect when people are passing inline functions, so we could warn.